cql3/restrictions/statement_restrictions.cc

```diff
@@ -553,14 +600,93 @@ void statement_restrictions::process_clustering_columns_restrictions(bool for_vi
+    if (values == expr::value_set(expr::value_list{})) {
+        return {};
+    }
+    const auto bounds = expr::to_range(values);
+    const auto start_token = bounds.start() ? dht::token::from_bytes(bounds.start()->value())
+            : dht::minimum_token();
+    auto end_token = bounds.end() ? dht::token::from_bytes(bounds.end()->value()) : dht::maximum_token();
+    const bool include_start = bounds.start() && bounds.start()->is_inclusive();
+    const auto include_end = bounds.end() && bounds.end()->is_inclusive();
+
+    auto start = dht::partition_range::bound(include_start
+                       ? dht::ring_position::starting_at(start_token)
+                       : dht::ring_position::ending_at(start_token));
+    auto end = dht::partition_range::bound(include_end
+                     ? dht::ring_position::ending_at(end_token)
+                     : dht::ring_position::starting_at(end_token));
+
+    return {{std::move(start), std::move(end)}};
+}
+
+/// Turns a partition-key value into a partition_range. \p pk must have elements for all partition columns.
+dht::partition_range range_from_bytes(const schema& schema, const std::vector<bytes>& pk) {
+    const auto k = partition_key::from_exploded(pk);
+    const query::ring_position pos(dht::get_token(schema, k), k);
+    return dht::partition_range::make_singular(std::move(pos));
+}
+
+void error_if_exceeds(size_t size, size_t limit) {
+    if (size > limit) {
+        throw std::runtime_error(
+                fmt::format("clustering-key cartesian product size {} is greater than maximum {}", size, limit));
+    }
+}
+
+/// Computes partition-key ranges from expressions, which contains EQ/IN for every partition column.
+dht::partition_range_vector partition_ranges_from_singles(
+        const std::vector<expr::expression>& expressions, const query_options& options, const schema& schema) {
+    const size_t size_limit =
+            options.get_cql_config().restrictions.partition_key_restrictions_max_cartesian_product_size;
+    // Each element is a vector of that column's possible values:
+    std::vector<std::vector<bytes>> column_values(schema.partition_key_size());
+    size_t product_size = 1;
+    for (const auto& e : expressions) {
+        if (const auto arbitrary_binop = find_atom(e, [] (const binary_operator&) { return true; })) {
+            if (auto cv = std::get_if<expr::column_value>(&arbitrary_binop->lhs)) {
+                const value_set vals = possible_lhs_values(cv->col, e, options);
+                if (auto lst = std::get_if<value_list>(&vals)) {
+                    if (lst->empty()) {
+                        return {};
+                    }
+                    product_size *= lst->size();
+                    error_if_exceeds(product_size, size_limit);
+                    column_values[schema.position(*cv->col)] = *lst;
+                } else {
+                    throw invalid_request("use token!");
```

Comment:
In what circumstances would users see this error message? Will there be more context? If no, then perhaps we should elaborate a little bit more to help the users fix their queries

Reply:
Oh, God, sorry.  This is a remnant of an early draft, it should be much friendlier.  Thanks for catching it.